File: examples_extra_prime/C#/ActivateLicense/ActivateLicense.cs
/****************************************************************************
**
** Copyright 2015 by Emotiv. All rights reserved
** Example - ActivateLicense
** How to activate a license key
 ***
****************************************************************************/

using System;
using System.Collections.Generic;
using Emotiv;
using System.IO;
using System.Threading;
using System.Reflection;

namespace ActivateLicense
{
    class ActivateLicense
    {
        static string licenseKey = "";           // Your License Key

        static string format = "dd/MM/yyyy";
        static Int32 debitNum = 2; //example 
        static int userCloudID = 0;
        static int userID = -1;
        static string userName = "your_UserName";
        static string password = "your_Password";
        static string profileName = "Profile_1";

        static void activateLicense()
        {
            int result = EdkDll.IEE_AuthorizeLicense(licenseKey, debitNum);
            if (result == EdkDll.EDK_OK || result == EdkDll.EDK_LICENSE_REGISTERED)
            {
                Console.WriteLine("License activated.");
            }
            else Console.WriteLine("License Error. " + result);
        }

        static public DateTime FromUnixTime(uint unixTime)
        {
            var epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
            return epoch.AddSeconds(unixTime);
        }

        static public long ToUnixTime(DateTime date)
        {
            var epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
            return Convert.ToInt64((date - epoch).TotalSeconds);
        }

        static void getDebitInformation()
        {
            EdkDll.IEE_DebitInfos_t debitInfos = new EdkDll.IEE_DebitInfos_t();
            int result = EdkDll.IEE_GetDebitInformation(licenseKey, ref debitInfos);

            Console.WriteLine();
            Console.WriteLine("Total debit a day            : " + debitInfos.total_debit_today);
            Console.WriteLine("Daily debit limit            : " + debitInfos.daily_debit_limit);
            Console.WriteLine("Remain Session               : " + debitInfos.remainingSessions);
            Console.WriteLine();
            Console.WriteLine("Remained time (seconds) before resetting debit limitation a day: " + debitInfos.time_reset);
            Console.WriteLine();
        }

        static void licenseInformation()
        {
            EdkDll.IEE_LicenseInfos_t licenseInfos = new EdkDll.IEE_LicenseInfos_t();
            int result = EdkDll.IEE_LicenseInformation(ref licenseInfos);

            Console.WriteLine();
            Console.WriteLine("Date From            : " + FromUnixTime(licenseInfos.date_from).ToString(format));
            Console.WriteLine("Date To              : " + FromUnixTime(licenseInfos.date_to).ToString(format));
            Console.WriteLine();

            Console.WriteLine();
            Console.WriteLine("Soft Limit Date      : " + FromUnixTime(licenseInfos.soft_limit_date).ToString(format));
            Console.WriteLine("Hard Limit Date      : " + FromUnixTime(licenseInfos.hard_limit_date).ToString(format));
            Console.WriteLine();

            Console.WriteLine("Seat number          : " + licenseInfos.seat_count);
            Console.WriteLine();

            Console.WriteLine("Total Quota          : " + licenseInfos.quota);
            Console.WriteLine("Total quota used     : " + licenseInfos.usedQuota);
            Console.WriteLine();

            switch ((int)licenseInfos.scopes)
            {
                case (int)EdkDll.IEE_LicenseType_t.IEE_EEG:

                    Console.WriteLine("License type : EEG");
                    Console.WriteLine();
                    break;
                case (int)EdkDll.IEE_LicenseType_t.IEE_EEG_PM:

                    Console.WriteLine("License type : EEG + PM");
                    Console.WriteLine();
                    break;
                case (int)EdkDll.IEE_LicenseType_t.IEE_PM:
                    Console.WriteLine("License type : PM");
                    Console.WriteLine();
                    break;
                default:
                    Console.WriteLine("License type : No type");
                    Console.WriteLine();
                    break;
            }
        }

        static void Main(string[] args)
        {
            Console.WriteLine("===========================================");
            Console.WriteLine("The example to activate a license key.");
            Console.WriteLine("===========================================");
            
            EmoEngine engine = EmoEngine.Instance;
            engine.Connect();

            //Authorize
            if (EmotivCloudClient.EC_Connect() != EdkDll.EDK_OK)
            {
                Console.WriteLine("Cannot connect to Emotiv Cloud.");
                Thread.Sleep(2000);
                return;
            }

            if (EmotivCloudClient.EC_Login(userName, password) != EdkDll.EDK_OK)
            {
                Console.WriteLine("Your login attempt has failed. The username or password may be incorrect");
                Thread.Sleep(2000);
                return;
            }

            Console.WriteLine("Logged in as " + userName);

            if (EmotivCloudClient.EC_GetUserDetail(ref userCloudID) != EdkDll.EDK_OK)
                return;

            //Active license
            activateLicense();

            //We can call this API any time to check current License information
            licenseInformation();

            //GetDebitInfo
            getDebitInformation();

            Thread.Sleep(5000);
        }
    }
}

File: examples_extra_prime/C++/ActivateLicense/ActivateLicense.cpp
/****************************************************************************
**
** Copyright 2017 by Emotiv. All rights reserved
** Example - ActivateLicense
** How to activate a license key and Get infomation of that license 
**
****************************************************************************/

#ifdef _WIN32
    #include <windows.h>
    #include <conio.h>
#endif

#if __linux__ || __APPLE__
#include <unistd.h>
#include <termios.h>
int _getch(void);
#endif

#include <iostream>
#include <sstream>
#include <time.h>
#include <stdio.h>
#include <ctime>

#include "Iedk.h"
#include "IedkErrorCode.h"
#include "EmotivCloudClient.h"

std::string convertEpochToTime(time_t epochTime, std::string format = "%Y-%m-%d %H:%M:%S");
std::string convertEpochToTime(time_t epochTime, std::string format)
{
    if (format == "")
        format = "%Y-%m-%d %H:%M:%S";


    char timestamp[64] = { 0 };
    strftime(timestamp, sizeof(timestamp), format.c_str(), localtime(&epochTime));
    return timestamp;
}

std::string intToHex(int x)
{
    std::stringstream stream;
    stream << std::hex << x;
    std::string result(stream.str());
    result = "0x" + result;

    return result;
}

void printLicenseInformation(IEE_LicenseInfos_t& licenseInfos)
{
    int licenseType = 0;

    std::cout << std::endl;
    std::cout << "From Date         : " << convertEpochToTime(licenseInfos.date_from) << std::endl;
    std::cout << "To   Date         : " << convertEpochToTime(licenseInfos.date_to) << std::endl;
    std::cout << std::endl;

    std::cout << std::endl;
    std::cout << "Soft Limit Date   : " << convertEpochToTime(licenseInfos.soft_limit_date) << std::endl;
    std::cout << "Hard Limit Date   : " << convertEpochToTime(licenseInfos.hard_limit_date) << std::endl;
    std::cout << std::endl;

    std::cout << "Number of Seats   : " << licenseInfos.seat_count << std::endl;
    std::cout << std::endl;

    std::cout << "Total Quota       : " << licenseInfos.quota << std::endl;
    std::cout << "Total used Quota  : " << licenseInfos.usedQuota << std::endl;
    std::cout << std::endl;

    switch (licenseInfos.scopes)
    {
    case IEE_EEG:
        licenseType = IEE_LicenseType_t::IEE_EEG;

        std::cout << "License type  : " << "EEG" << std::endl;
        std::cout << std::endl;
        break;
    case IEE_EEG_PM:
        licenseType = IEE_LicenseType_t::IEE_EEG_PM;

        std::cout << "License type  : " << "EEG + PM" << std::endl;
        std::cout << std::endl;
        break;
    case IEE_PM:
        licenseType = IEE_LicenseType_t::IEE_PM;
        std::cout << "License type  : " << "PM" << std::endl;
        std::cout << std::endl;
        break;
    default:
        std::cout << "License type  : " << "No type" << std::endl;
        std::cout << std::endl;
        break;
    }
}

std::string const LICENSE_KEY = "";
                                 

int main(int argc, char** argv)
{

    //Login request
    std::string userName = "";
    std::string password = "";
    int result = 0;

    if (IEE_EngineConnect() != EDK_OK) {
        std::cout << "Emotiv Driver start up failed.";
        return -1;
    }

    if (EC_Connect() != EDK_OK)
    {
        std::cout << "Cannot connect to Emotiv Cloud";
        return -1;
    }
    //Login
    if (EC_Login(userName.c_str(), password.c_str()) != EDK_OK)
    {
        std::cout << "Login failed. The username or password may be incorrect" << std::endl;
        return -1;
    }

    //get Debit info
    IEE_DebitInfos_t debitInfos;
    debitInfos.remainingSessions = 0;
    result = IEE_GetDebitInformation(LICENSE_KEY.c_str(), &debitInfos);

    if (debitInfos.total_session_inYear == 0)
    {
        std::cout << std::endl;
        std::cout << "Remaining Sessions in month : " << debitInfos.remainingSessions << std::endl;
        std::cout << "Daily debit limitation      : " << debitInfos.daily_debit_limit << std::endl;
        std::cout << std::endl;

        std::cout << std::endl;
        std::cout << "Total debit today           : " << debitInfos.total_debit_today << std::endl;
        std::cout << "Remaining time before resetting daily debit limitation: " << debitInfos.time_reset << "(seconds)" << std::endl;
        std::cout << std::endl;
    }    
    else {
        std::cout << std::endl;
        std::cout << "Remaining Sessions in year : " << debitInfos.remainingSessions << std::endl;
        std::cout << "the total number of session can be debitable in year : " << debitInfos.total_session_inYear << std::endl;
        std::cout << std::endl;
    }


    //Active license with debit
    unsigned int debitNum = 0; //default value

    //Get number of debit as input
    std::string input;
    std::cout << "Please give number of debits : " << std::endl;

    std::getline(std::cin, input, '\n');
    debitNum = atoi(input.c_str());

    result = IEE_AuthorizeLicense(LICENSE_KEY.c_str(), debitNum);

    //show result after authorizing
    switch (result)
    {
    case EDK_INVALID_DEBIT_NUMBER:
    case EDK_INVALID_DEBIT_ERROR:
        std::cout << "Invalid number of Debits" << std::endl;
        break;
    case EDK_INVALID_PARAMETER:
        std::cout << "Invalid user information" << std::endl;
        break;
    case EDK_NO_INTERNET_CONNECTION:
        std::cout << "No Internet Connection" << std::endl;
        break;
    case EDK_LICENSE_EXPIRED:
        std::cout << "Expired license" << std::endl;
        break;
    case EDK_OVER_DEVICE_LIST:
        std::cout << "Over device list" << std::endl;
        break;  
    case EDK_DAILY_DEBIT_LIMITED:
        std::cout << "Over daily number of debits" << std::endl;
        break;
    case EDK_ACCESS_DENIED:
        std::cout << "Access denied" << std::endl;
        break;
    case EDK_LICENSE_REGISTERED:
        std::cout << "The License has registered" << std::endl;
        break;
    case EDK_LICENSE_ERROR:
        std::cout << "Error License" << std::endl;
        break;
    case EDK_LICENSE_NOT_FOUND:
        std::cout << "License not found" << std::endl;
        break;
    case EDK_UNKNOWN_ERROR:
        break;
    default:
        break;
    }

    if (!(result == EDK_OK || result == EDK_LICENSE_REGISTERED))
        return result;
    
    IEE_LicenseInfos_t licenseInfos;

    // We can call this API any time to check current License information
    result = IEE_LicenseInformation(&licenseInfos);

    std::cout << std::endl << "IEE_LicenseInfomation result = " << intToHex(result) << std::endl;

    switch (result)
    {
        case EDK_LICENSE_EXPIRED:
            std::cout << "Expired license" << std::endl;
            break;
        case EDK_OVER_QUOTA:
            std::cout << "Over quota" << std::endl;
            break;
        case EDK_ACCESS_DENIED:
            std::cout << "Access denined" << std::endl;
            break;
        case EDK_LICENSE_ERROR:
            std::cout << "Error License" << std::endl;
            break;
        case EDK_NO_ACTIVE_LICENSE:
            std::cout << "No active license" << std::endl;
            break;
        default:
            break;
    }
    printLicenseInformation(licenseInfos);

    _getch();

}
#ifdef __APPLE__
int _getch(void)
{
    int r;
    unsigned char c;
    if ((r = read(0, &c, sizeof(c))) < 0)
    {
        return r;
    }
    else
    {
        return c;
    }
}
#endif
